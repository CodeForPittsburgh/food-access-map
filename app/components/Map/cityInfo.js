import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export default class CityInfo extends React.PureComponent {
  render() {
    const { info } = this.props;

    return (
      <div>
        <Link to={`/foodSite/${info.id}`}>
          Go to detail page for {info.name}
        </Link>
      </div>
    );
  }
}

CityInfo.propTypes = {
  info: PropTypes.object.isRequired,
};
