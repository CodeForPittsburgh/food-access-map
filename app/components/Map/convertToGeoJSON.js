// Converts api data to geoJSON points
// Might be better if this arrived from the server in this format

const convertToGeoJSON = ({ data }) => ({
  type: 'FeatureCollection',
  features: data.reduce((result, site) => {
    const { latitude, longitude } = site;
    if (longitude) {
      result.push({
        type: 'Feature',
        geometry: {
          type: 'Point',
          // GeoJSON takes lat/lon in reverse order
          coordinates: [longitude, latitude],
        },
        properties: site,
      });
    }
    return result;
  }, []),
});

export default convertToGeoJSON;
