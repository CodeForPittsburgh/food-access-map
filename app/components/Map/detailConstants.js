export const detailArray = ['all', 'wic', 'snap'];
