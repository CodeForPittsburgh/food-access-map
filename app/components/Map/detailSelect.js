import React, { Component } from 'react';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core';

const styles = theme => ({
  root: {
    position: 'absolute',
    top: theme.spacing(16),
    left: theme.spacing(4),
    zIndex: 100,
    background: '#FAFAFA',
  },
});

/* eslint-disable react/prefer-stateless-function */
class DetailSelect extends Component {
  render() {
    const details = this.props.detailArray;
    const { classes } = this.props;

    return (
      <div className={classes.root}>
        <FormControl>
          <InputLabel htmlFor="detail-selector">Details</InputLabel>
          <Select
            value={this.props.selectedDetail}
            onChange={this.props.handleChange}
            input={<Input name="detailSelector" id="detail-selector" />}
          >
            {details.map(detail => (
              <MenuItem key={detail} value={detail}>
                {detail.toUpperCase()}
              </MenuItem>
            ))}
          </Select>
          <FormHelperText>Filter by important details</FormHelperText>
        </FormControl>
      </div>
    );
  }
}

DetailSelect.propTypes = {
  detailArray: PropTypes.array.isRequired,
  classes: PropTypes.object,
  selectedDetail: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
};

export default withStyles(styles)(DetailSelect);
