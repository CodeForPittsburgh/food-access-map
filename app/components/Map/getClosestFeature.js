// It's good to have a large click radius for mobile to be forgiving with touch precision
// However it is worthwhile to calculate the closest feature in case there are several
// features within the click radius and you are trying to be precise.

const getClosestFeature = event => {
  const [pointerLng, pointerLat] = event.lngLat;
  const first = event.features.shift();
  const closest = {
    offset:
      Math.abs(pointerLng - first.longitude) +
      Math.abs(pointerLat - first.latitude),
    properties: first.properties,
  };
  event.features.forEach(({ properties }) => {
    const currOffset =
      Math.abs(pointerLng - properties.longitude) +
      Math.abs(pointerLat - properties.latitude);
    if (currOffset < closest.offset) {
      closest.offset = currOffset;
      closest.properties = properties;
    }
  });
  return closest.properties;
};

export default getClosestFeature;
