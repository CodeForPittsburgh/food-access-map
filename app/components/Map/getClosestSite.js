// It's good to have a large click radius for mobile to be forgiving with touch precision
// However it is worthwhile to calculate the closest feature in case there are several
// features within the click radius and you are trying to be precise.

const getClosestSite = (features, pointerCoords) => {
  const [pointerLng, pointerLat] = pointerCoords;
  const first = features.shift();
  const closest = {
    offset:
      Math.abs(pointerLng - first.properties.longitude) +
      Math.abs(pointerLat - first.properties.latitude),
    properties: first.properties,
  };
  features.forEach(({ properties }) => {
    const currOffset =
      Math.abs(pointerLng - properties.longitude) +
      Math.abs(pointerLat - properties.latitude);
    if (currOffset < closest.offset) {
      closest.offset = currOffset;
      closest.properties = properties;
    }
  });
  return closest.properties;
};

export default getClosestSite;
