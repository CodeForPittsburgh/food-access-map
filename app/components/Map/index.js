/**
 *
 * Map
 *
 */

import React, { memo } from 'react';
// import PropTypes from 'prop-types';
// import styled from 'styled-components';
import _ from 'lodash';

import ReactMapGL from 'react-map-gl';
import MapSelect from './mapSelect';
import { townArray } from './townConstants';

// import { FormattedMessage } from 'react-intl';
// import messages from './messages';

/* eslint-disable react/prefer-stateless-function */
class Map extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      viewport: {
        width: 400,
        height: 400,
        latitude: 40.4406,
        longitude: -79.9659,
        zoom: 8,
      },
      selectedTown: 'Pittsburgh',
    };
  }

  handleSelection(event) {
    const oldView = _.clone(this.state.viewport);
    const place = event.target.value;
    const { latitude, longitude } = _.find(townArray, { place });
    const newView = _.assign({}, oldView, {
      zoom: 12,
      latitude,
      longitude,
    });

    this.setState({
      viewport: newView,
      selectedTown: place,
    });
  }

  render() {
    return (
      <div>
        <ReactMapGL
          {...this.state.viewport}
          onViewportChange={viewport => this.setState({ viewport })}
          mapboxApiAccessToken="pk.eyJ1IjoiaHlwZXJmbHVpZCIsImEiOiJjaWpra3Q0MnIwMzRhdGZtNXAwMzRmNXhvIn0.tZzUmF9nGk2h28zx6PM13w"
        />
        <MapSelect
          townArray={townArray}
          selectedTown={this.state.selectedTown}
          handleChange={e => this.handleSelection(e)}
        />
      </div>
    );
  }
}

Map.propTypes = {};

export default memo(Map);
