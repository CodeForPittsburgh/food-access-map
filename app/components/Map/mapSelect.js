import React, { Component } from 'react';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import PropTypes from 'prop-types';

/* eslint-disable react/prefer-stateless-function */
class MapSelect extends Component {
  render() {
    const towns = this.props.townArray;

    return (
      <div>
        <FormControl>
          <InputLabel htmlFor="town-selector">Town/Neighborhood</InputLabel>
          <Select
            value={this.props.selectedTown}
            onChange={this.props.handleChange}
            input={<Input name="townSelector" id="town-selector" />}
          >
            {towns.map(town => (
              <MenuItem key={town.latitude + town.longitude} value={town.place}>
                {town.place}
              </MenuItem>
            ))}
          </Select>
          <FormHelperText>Some important helper text</FormHelperText>
        </FormControl>
      </div>
    );
  }
}

MapSelect.propTypes = {
  townArray: PropTypes.array.isRequired,
  selectedTown: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
};

export default MapSelect;
