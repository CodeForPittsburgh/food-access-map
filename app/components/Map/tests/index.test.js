/**
 *
 * Tests for Map
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import Map from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';

configure({ adapter: new Adapter() });
describe('<Map />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Map />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('should call set state when there are changes', () => {
    const wrapper = shallow(<Map />);
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'setState');
    instance.handleSelection({ target: { value: 'Millvale' } });
    expect(spy).toHaveBeenCalledWith({
      viewport: {
        width: 400,
        height: 400,
        latitude: 40.4801,
        longitude: -79.9784,
        zoom: 12,
      },
      selectedTown: 'Millvale',
    });
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Map />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
