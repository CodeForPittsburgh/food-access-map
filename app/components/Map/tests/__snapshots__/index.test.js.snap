// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> Should render and match the snapshot 1`] = `
<div>
  <div
    style="position: relative; width: 0px; height: 0px; cursor: grab; outline: none;"
    tabindex="1"
  >
    <div
      style="position: relative; width: 100%; height: 100%;"
    >
      <div
        class=""
        style="position: absolute; width: 100%; height: 100%; overflow: hidden; visibility: inherit;"
      />
      <div
        style="overflow: visible; height: 0px; width: 0px;"
      />
      <div
        class="resize-triggers"
      >
        <div
          class="expand-trigger"
        >
          <div
            style="width: 1px; height: 1px;"
          />
        </div>
        <div
          class="contract-trigger"
        />
      </div>
    </div>
  </div>
  <div>
    <div
      class="MuiFormControl-root-81"
    >
      <label
        class="MuiFormLabel-root-96 MuiFormLabel-filled-100 MuiInputLabel-root-85 MuiInputLabel-formControl-90 MuiInputLabel-animated-93 MuiInputLabel-shrink-92"
        data-shrink="true"
        for="town-selector"
      >
        Town/Neighborhood
      </label>
      <div
        class="MuiInputBase-root-123 MuiInput-root-110 MuiInput-underline-114 MuiInputBase-formControl-124 MuiInput-formControl-111"
      >
        <div
          class="MuiSelect-root-103"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            class="MuiSelect-select-104 MuiSelect-selectMenu-107 MuiInputBase-input-133 MuiInput-input-118"
            id="select-townSelector"
            role="button"
            tabindex="0"
          >
            Pittsburgh
          </div>
          <input
            id="town-selector"
            name="townSelector"
            type="hidden"
            value="Pittsburgh"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-140 MuiSelect-icon-109"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <p
        class="MuiFormHelperText-root-153 MuiFormHelperText-filled-159"
      >
        Choose a region for instant zoom
      </p>
    </div>
  </div>
</div>
`;
